crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1,91 +1,70 @@
 use ruff_db::files::File;
-use ruff_db::parsed::parsed_module;
 use ruff_python_ast::name::Name;
 
-use crate::semantic_index::symbol::{NodeWithScopeKind, PublicSymbolId, ScopeId};
-use crate::semantic_index::{public_symbol, root_scope, semantic_index, symbol_table};
-use crate::types::infer::{TypeInference, TypeInferenceBuilder};
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{ScopeId, ScopedSymbolId};
+use crate::semantic_index::{root_scope, symbol_table, use_def_map};
 use crate::{Db, FxOrderSet};
 
 mod display;
 mod infer;
 
-/// Infers the type of a public symbol.
-///
-/// This is a Salsa query to get symbol-level invalidation instead of file-level dependency invalidation.
-/// Without this being a query, changing any public type of a module would invalidate the type inference
-/// for the module scope of its dependents and the transitive dependents because.
-///
-/// For example if we have
-/// ```python
-/// # a.py
-/// import x from b
-///
-/// # b.py
-///
-/// x = 20
-/// ```
-///
-/// And x is now changed from `x = 20` to `x = 30`. The following happens:
-///
-/// * The module level types of `b.py` change because `x` now is a `Literal[30]`.
-/// * The module level types of `a.py` change because the imported symbol `x` now has a `Literal[30]` type
-/// * The module level types of any dependents of `a.py` change because the imported symbol `x` now has a `Literal[30]` type
-/// * And so on for all transitive dependencies.
-///
-/// This being a query ensures that the invalidation short-circuits if the type of this symbol didn't change.
-#[salsa::tracked]
-pub(crate) fn public_symbol_ty<'db>(db: &'db dyn Db, symbol: PublicSymbolId<'db>) -> Type<'db> {
-    let _span = tracing::trace_span!("public_symbol_ty", ?symbol).entered();
-
-    let file = symbol.file(db);
-    let scope = root_scope(db, file);
-
-    // TODO switch to inferring just the definition(s), not the whole scope
-    let inference = infer_types(db, scope);
-    inference.symbol_ty(symbol.scoped_symbol_id(db))
-}
+pub(crate) use self::infer::{infer_definition_types, infer_expression_types, infer_scope_types};
 
-/// Shorthand for `public_symbol_ty` that takes a symbol name instead of a [`PublicSymbolId`].
-pub(crate) fn public_symbol_ty_by_name<'db>(
+/// Infer the public type of a symbol (its type as seen from outside its scope).
+pub(crate) fn symbol_ty<'db>(
```

Comment:
Is idea here that we don't cache the public symbol type anymore because we already cache the definition types?

Reply:
Yeah... the thing is that all symbols (not just root-scope symbols) have a "public type" (i.e. the type as seen from another scope) -- for non-root-scope symbols they can be seen by a nested scope, for class-scope symbols they can also be seen by other scopes via class attribute access -- and I want to handle all of them the same way. And I'm not sure we want every symbol to be an ingredient.

But this will be fairly easy to change in the future, I think; it can be an empirical question whether performance is better if we make all symbols an ingredient and cache public type of all symbols, or don't make them an ingredient and sometimes re-create the same union.